keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+  // resource to be created if set. The name of this StatefulApplication resource will be used if this is
+  // not set. It’s assumed that the Application resource is in the same namespace of it exists.
+  // +optional
+  ApplicationName *string `json:"applicationName,omitempty"`
+  
+  // Source specifies an optional data source from which the application and its data state is restored.
+  // +optional
+  Source *DataSource `json:"source,omitempty"`
+
+  // ValueSubstitutionRules is a set of rules for value substitution when the application is restored
+	// from a data source.
+	// +optional
+	ValueSubstitutionRules []ValueSubstitutionRule `json:"valueSubstitutionRules,omitempty"`
+}
+
+type StatefulApplicationStatus struct {
+  // Conditions show the current condition of this application.
+  Conditions []ApplicationCondition `json:"conditions,omitempty"`
+  
+  // Source is a reference to the source from which the application was most recently restored or rolled back from.
+  Source *corev1.TypedLocalObjectReference `json:"source,omitempty"`
+}
+
+// ValueSubstitutionRuleType is the type of ValueSubstitutionRules.
+type ValueSubstitutionRuleType string
+
+// Set of valid values of ValueSubstitutionRuleType.
+const (
+	ValueSubstitutionRuleTypeName       ValueSubstitutionRuleType = "Name"
+	ValueSubstitutionRuleTypeLabel      ValueSubstitutionRuleType = "Label"
+	ValueSubstitutionRuleTypeAnnotation ValueSubstitutionRuleType = "Annotation"
+	ValueSubstitutionRuleTypeEnvVar     ValueSubstitutionRuleType = "EnvVar"
+)
+
+// ValueSubstitutionRule defines a rule for value sibstitution.
+type ValueSubstitutionRule struct {
+	// Type is the type of this ValueSubstitutionRule.
+	Type ValueSubstitutionRuleType `json:"type"`
+
+	// Key is an optional key this rule applies to. For example, when the type is Label, this can be used to
+	// specify the label this rule applies to. Similarly, when the type is Annotation or EnvVar, this can be
+	// used to specify the annotation or environment variable this rule applies to.
+	// +optional
+	Key *string `json:"key,omitempty"`
+
+	// OldValue is the old value to be substituted in the given type of values.
+	// This may be a pattern in the values.
+	OldValue string `json:"oldValue"`
+
+	// NewValue is the new value that will be used in place of each occurrence of the OldValue.
+	// If NewValue is empty, it is considered an intention to remove the key with the OldValue.
+	NewValue string `json:"newValue"`
+
+	// Selector is an optional label selector for filtering Kubernetes resources to which this rule applies.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+// DataSourceType specifies the type of the data source from which the application and its data state is restored.
+type DataSourceType string
+
+// Set of valid values of DataSourceType.
+const (
+  DataSourceSnapshot DataSourceType = "Snapshot"
+  DataSourceBackup   DataSourceType = "Backup"
+)
```

Comment:
There is work going on to support volume cloning. A data source of "PVC" will be possible for creating a PVC when that work is complete.  So here you may have a third type ("Application"?) in the future.

Reply:
Both Snapshot and PVC DataSource are in beta now.
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L426